File: script/engine/animating/sampler.ts
import { quat } from "../core/math/quat.js";
import { vec3, Vec3Like } from "../core/math/vec3.js";
import { Vec4, vec4 } from "../core/math/vec4.js";

const vec4_a = vec4.create();
const vec4_b = vec4.create();

function binarySearch(source: ArrayLike<number>, value: number, EPSILON = 1e-6): number {
    if (value < source[0]) {
        return 0;
    }

    if (value > source[source.length - 1]) {
        return source.length - 1;
    }

    let head = 0;
    let tail = source.length - 1;
    while (head <= tail) {
        const mid = (head + tail) >>> 1;
        const res = source[mid];
        if ((value + EPSILON) < res) {
            tail = mid - 1;
        } else if ((value - EPSILON) > res) {
            head = mid + 1;
        } else {
            return mid;
        }
    }
    return ~head;
}

export function sampleVec3(out: Vec3Like, input: ArrayLike<number>, output: ArrayLike<number>, time: number) {
    const index = binarySearch(input, time);
    if (index >= 0) {
        const start = index * 3;
        vec3.set(out, output[start], output[start + 1], output[start + 2]);
    } else {
        const next = ~index;
        const prev = next - 1;

        const start_prev = prev * 3;
        vec3.set(vec4_a, output[start_prev], output[start_prev + 1], output[start_prev + 2])

        const start_next = next * 3;
        vec3.set(vec4_b, output[start_next], output[start_next + 1], output[start_next + 2])

        const t = (time - input[prev]) / (input[next] - input[prev]);
        vec3.lerp(out, vec4_a, vec4_b, t);
    }
}

export function sampleQuat(out: Vec4, input: ArrayLike<number>, output: ArrayLike<number>, time: number) {
    const index = binarySearch(input, time);
    if (index >= 0) {
        const start = index * 4;
        quat.set(out, output[start], output[start + 1], output[start + 2], output[start + 3]);
    } else {
        const next = ~index;
        const prev = next - 1;

        const start_prev = prev * 4;
        quat.set(vec4_a, output[start_prev], output[start_prev + 1], output[start_prev + 2], output[start_prev + 3])

        const start_next = next * 4;
        quat.set(vec4_b, output[start_next], output[start_next + 1], output[start_next + 2], output[start_next + 3])

        const t = (time - input[prev]) / (input[next] - input[prev]);
        quat.slerp(out, vec4_a, vec4_b, t);
    }
}

File: puttyknife/sampler/sampler.cpp
#include <sampler/sampler.hpp>

#define EPSILON 1e-6f

int sampler_binarySearch(float *srcData, unsigned int srcLength, float value)
{
    if (value < srcData[0])
    {
        return 0;
    }

    if (value > srcData[srcLength - 1])
    {
        return srcLength - 1;
    }

    unsigned int head = 0;
    unsigned int tail = srcLength - 1;
    while (head <= tail)
    {
        unsigned int mid = (head + tail) >> 1;
        float res = srcData[mid];
        if ((value + EPSILON) < res)
        {
            tail = mid - 1;
        }
        else if ((value - EPSILON) > res)
        {
            head = mid + 1;
        }
        else
        {
            return mid;
        }
    }
    return ~head;
}

void sampler_vec3(float3 *out, float *inputData, unsigned int inputLength, float3 *output, float time)
{
    int index = sampler_binarySearch(inputData, inputLength, time);
    if (index >= 0)
    {
        *out = *(output + index);
    }
    else
    {
        int next = ~index;
        int prev = next - 1;

        float t = (time - inputData[prev]) / (inputData[next] - inputData[prev]);
        lerp(*out, *(output + prev), *(output + next), t);
    }
}

void sampler_quat(float4 *out, float *inputData, unsigned int inputLength, float4 *output, float time)
{
    int index = sampler_binarySearch(inputData, inputLength, time);
    if (index >= 0)
    {
        *out = *(output + index);
    }
    else
    {
        int next = ~index;
        int prev = next - 1;

        float t = (time - inputData[prev]) / (inputData[next] - inputData[prev]);
        slerp(*out, *(output + prev), *(output + next), t);
    }
}